okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java

```diff
@@ -0,0 +1,132 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.squareup.okhttp;
+
+import com.google.common.base.Preconditions;
+
+import com.squareup.okhttp.internal.Platform;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+
+/**
+ * A helper class that located in package com.squareup.okhttp, so that we can use OkHttp internals
+ * to do TLS upgrading.
+ */
+public final class OkHttpTlsUpgrader {
+
+  // A dummy address used to bypass null check.
+  private static final InetSocketAddress DUMMY_INET_SOCKET_ADDRESS =
+      InetSocketAddress.createUnresolved("fake", 73);
+  public static final ConnectionSpec DEFAULT_SPEC =
+      new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
+      .cipherSuites(
+          // The following items should be sync with Netty's Http2SecurityUtil.CIPHERS.
+          CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+          CipherSuite.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+          CipherSuite.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+          CipherSuite.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+          CipherSuite.TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
+          CipherSuite.TLS_DHE_DSS_WITH_AES_128_GCM_SHA256,
+          CipherSuite.TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,
+          CipherSuite.TLS_DHE_DSS_WITH_AES_256_GCM_SHA384)
+      .tlsVersions(TlsVersion.TLS_1_2)
+      .supportsTlsExtensions(true)
+      .build();
+
+
+  /**
+   * Upgrades given Socket to be a SSLSocket.
+   */
+  public static SSLSocket upgrade(SSLSocketFactory sslSocketFactory,
+      Socket socket, String host, int port, @Nullable ConnectionSpec spec) throws IOException {
+    SSLSocket sslSocket = (SSLSocket) sslSocketFactory.createSocket(
+        socket, host, port, true /* auto close */);
+    if (spec == null) {
+      spec = DEFAULT_SPEC;
+    }
+    spec.apply(sslSocket, getOkHttpRoute(host, port, spec));
+
+    Platform platform = Platform.get();
+    try {
+      // Force handshake.
+      sslSocket.startHandshake();
+
+      String negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
+      if (negotiatedProtocol == null) {
+        throw new RuntimeException("protocol negotiation failed");
+      }
+      Preconditions.checkState(Protocol.HTTP_2.equals(Protocol.get(negotiatedProtocol)),
+          "negotiated protocol is " + negotiatedProtocol + " instead of http2.");
```

Comment:
We don't want to do the string concatenation unless the protocol is wrong:
```
"negotiated protocol is %s instead of http2.", negotiatedProtocol);
```